### ISSUE TITLE
Bug causes critical data loss

### ISSUE THREAD

**Issue by drupallerina** [docksal/docksal#191]
**Description**

`rm -rf` causes home folder on host to be deleted during `fin cleanup`.

**Steps to reproduce the issue:**
I recommend you do NOT try and reproduce this ;)

Background: I don't know whether this is relevant, but I moved my docker images from their default location on var/lib to a custom folder on a different partition following the docker documentation here (https://wiki.archlinux.org/index.php/Docker#Images_location)

1. Create commands/init containing `fin reset -f`
2. Start fin environment `fin up`
3. Re-initialise using `fin init`

**Describe the results you received:**

Containers were removed. So was my entire $HOME directory. (whelp!)

**Describe the results you expected:**

Containers (not even volumes or images) to be removed, like I confirmed in the command prompt:

> Preparing to delete all currently stopped containers

**Comment by achekulaev** [comment]
Hi

This is confusing at the moment.

1. `fin reset` does not do `fin cleanup`

```
11:31:14 Alexei-Workstation /Users/alexei.chekulaev/Projects/demo2
$ fin reset -f
Removing containers...
Removing network demo2_default
WARNING: Network demo2_default not found.
Removing volume demo2_project_root
WARNING: Volume demo2_project_root not found.
Volume docksal_ssh_agent is external, skipping
Removing volume demo2_host_home
WARNING: Volume demo2_host_home not found.
Starting services...
Creating network "demo2_default" with the default driver
Creating volume "demo2_project_root" with local driver
Creating volume "demo2_host_home" with local driver
Creating demo2_db_1
Creating demo2_cli_1
Creating demo2_web_1
Changing uid/gid in cli to 501/20 to match the host
Resetting permissions on /var/www...
Restarting php daemon...
Connected vhost-proxy to "demo2_default" network.
```

2. `fin cleanup` does not remove all stopped containers. Only `fin cleanup --hard` does. Apparently whoever created that `init` script used `--hard` switch (which he should have not used).

```
$ fin cleanup
Checking for orphaned containers...
Removing dangling images...
Removing dangling volumes...
defb402ae0d620358b81102e1941163896201a96b8200629be7e1a1d4f13f463
```

```
$ fin cleanup --hard
WARNING: Preparing to delete all currently stopped containers:
demo_web		Exited (0) 7 days ago (docksal/web:1.0-apache2.2)
demo_cli		Exited (137) 7 days ago (docksal/cli:1.2-php7)
demo_db		Exited (0) 7 days ago (docksal/db:1.0-mysql-5.5)
demo_mail		Exited (2) 7 days ago (mailhog/mailhog)
––––––––––––––––––––––––––––––––––––––––
Continue? [y/n]: y

Removing stopped containers...
5f38722ad404
aad181411c4e
0afd45f4c943
c1b04516cff2
Checking for orphaned containers...
Removing dangling images...
Removing dangling volumes...
demo_host_home
demo_project_root
```

3. Even `--hard` that I executed above did not remove my home folder. It's still there, all fine.

4. `fin init` is **your custom script** which we know nothing about. 

5. Please provide all the information new issue requested you to provide: fin sysinfo, fin config... 

6. Also provide contents of your `init` command

Thanks!

**Comment by drupallerina** [comment]
Thanks for the reply!
You're right, it can't be that line, it happened for a second time, after switching to master branch.
1. Ran fin init in my project directory (a Drupal 8 website)
2. Git pull recent changes
3. Run drush cim --> config import failed complaining about site uuids not matching
4. deleted settings.php and settings.local.php
5. fin init
6. watch my folders starting to disappear again: starting with the directory I'm in, going upwards in the tree, so maybe upfind has something to do with it?

## Content of fin init
```
#!/usr/bin/env bash

## Initialize Docksal powered Drupal 8 site
##
## Usage: fin init

# Abort if anything fails
set -e

#-------------------------- Settings --------------------------------

# PROJECT_ROOT is passed from fin.
# The following variables are configured in the '.env' file: DOCROOT, SITE_DIRECTORY, SOURCE_ALIAS and others.

DOCROOT_PATH="${PROJECT_ROOT}/${DOCROOT}"
SITEDIR_PATH="${DOCROOT_PATH}/sites/${SITE_DIRECTORY}"
#-------------------------- END: Settings --------------------------------

#-------------------------- Helper functions --------------------------------

# Console colors
red='\033[0;31m'
green='\033[0;32m'
green_bg='\033[42m'
yellow='\033[1;33m'
NC='\033[0m'

echo-red () { echo -e "${red}$1${NC}"; }
echo-green () { echo -e "${green}$1${NC}"; }
echo-green-bg () { echo -e "${green_bg}$1${NC}"; }
echo-yellow () { echo -e "${yellow}$1${NC}"; }

if_failed ()
{
	if [ ! $? -eq 0 ]; then
		if [[ "$1" == "" ]]; then msg="an error occurred"; else msg="$1"; fi
		echo-red "$msg";
		exit 1;
	fi
}

is_windows ()
{
	local res=$(uname | grep 'CYGWIN_NT')
	if [[ "$res" != "" ]]; then
		return 0
	else
		return 1
	fi
}

# Copy a settings file.
# Skips if the destination file already exists.
# @param $1 source file
# @param $2 destination file
copy_settings_file()
{
	local source="$1"
	local dest="$2"

	if [[ ! -f $dest ]]; then
		echo "Copying ${dest}..."
		cp $source $dest
	else
		echo-yellow "${dest} already in place."
	fi
}

#-------------------------- END: Helper functions --------------------------------

#-------------------------- Functions --------------------------------

# Initialize local settings files
init_settings ()
{

	# Copy from settings templates
	copy_settings_file "${PROJECT_ROOT}/config/default/default.settings.local.php" "${SITEDIR_PATH}/settings.local.php"
	copy_settings_file "${PROJECT_ROOT}/config/default/default.settings.php" "${SITEDIR_PATH}/settings.php"
	copy_settings_file "${PROJECT_ROOT}/config/default/development.services.yml" "${SITEDIR_PATH}/development.services.yml"

	# SETUP ONCE DURING _fin initing a cloned_ PROJECT
	# These are being 'compiled'
	if [[ ! -f ${SITEDIR_PATH}/drushrc.php ]]; then
	    envsubst '${VIRTUAL_HOST}' < ${PROJECT_ROOT}/config/default/drushrc.php >> ${SITEDIR_PATH}/drushrc.php
	fi

}

compile_configs ()
{
	# .gitlab-ci hooks
	if [[ ! -f ${PROJECT_ROOT}/.gitlab-ci.yml ]]; then
	    envsubst '${PROJECT_ID}' < ${PROJECT_ROOT}/config/default/.gitlab-ci.yml >> ${PROJECT_ROOT}/.gitlab-ci.yml
	fi


	# composer.json.
	if [[ ! -f ${PROJECT_ROOT}/composer.json ]]; then
	    envsubst '${PROJECT_ID} ${PROJECT_DESC}' < ${PROJECT_ROOT}/config/default/composer.json >> ${PROJECT_ROOT}/composer.json
	fi


	# README.md.
	if [[ ! -f ${PROJECT_ROOT}/README.md ]]; then
		envsubst '${GIT_URL} ${PROJECT_ID} ${VIRTUAL_HOST}' < ${PROJECT_ROOT}/config/default/README.md >> ${PROJECT_ROOT}/README.md
	fi

}

# Set file/folder permissions
file_permissions ()
{
	echo-green "Resetting files directory permissions..."
	mkdir -p "${SITEDIR_PATH}/files"
	chmod -R 777 "${SITEDIR_PATH}/files"
}

# Install site
site_install ()
{
	cd ${DOCROOT_PATH}
	fin drush si -y --site-name="${PROJECT_DESC}" --root=/var/www/web

	# Revert site dir permissions.
	chmod 755 "${SITEDIR_PATH}"
}

#-------------------------- END: Functions --------------------------------

#-------------------------- Execution --------------------------------

if [[ "$PROJECT_ROOT" == "" ]]; then
	echo-red "\$PROJECT_ROOT is not set"
	exit 1
fi

#echo-green "Setting file/folder permissions..."
#file_permissions

cd ${PROJECT_ROOT}

if [[ $DOCKER_RUNNING == "true" ]]; then
	echo -e "${green_bg} Step 1 ${NC}${green} Recreating services ...${NC}"
	fin reset -f
else
	echo -e "${green_bg} Step 1 ${NC}${green} Creating services ...${NC}"
	fin up
fi


echo -e "${green_bg} Step 2 ${NC}${green} Compiling root Config Files & composer install ...${NC}"
compile_configs

# Shouldnt be required anymore
# SETUP ONCE DURING _fin initing a cloned_ PROJECT
#if ! [ -s ${DOCROOT_PATH}/index.php ]; then
#	echo -e "${green_bg} Step 2b ${NC}${green} Drupal not found, downloading ...${NC}"
	#which is the correct way? the fin command says it cant autoload and cant run drupal-scaffold script	
#	fin exec composer "drupal-scaffold"
#	composer drupal-scaffold
#fi

composer install --ignore-platform-reqs


echo -e "${green_bg} Step 3 ${NC}${green} Initializing local project configuration ...${NC}"
init_settings

echo "Waiting 10s for MySQL to start...";
sleep 10


echo -e "${green_bg} Step 4 ${NC}${green} Installing site & updating drupal console ...${NC}"
time site_install

if [[ is_windows ]]; then
	echo-green "Add ${VIRTUAL_HOST} to your hosts file (/etc/hosts), e.g.:"
	echo-green "192.168.64.100  ${VIRTUAL_HOST}"
fi

fin cr

echo -e "${green_bg} Step 4 ${NC}${yellow} Drupal Console needs to be updated like this for now ...${NC}"
# Needed for some time because of a bug in docksal that didnt have the latest drupal console version
cd ${PROJECT_ROOT}
fin exec "sudo drupal self-update -y"


# SETUP ONCE DURING _creating a new_ PROJECT
# This is for our current configuration workflow to work
echo -e "${green_bg} Step 5 ${NC}${green} Setup SITE_UUID ...${NC}"
cd ${DOCROOT_PATH}
if [[ -z "$SITE_UUID"  ]]; then
    echo-yellow "SITE_UUID is not yet set and will be appended to .docksal/docksal.env"
    echo-yellow "This is fine when creating a Project from scratch but NOT if you CLONED IT!"
    export SITE_UUID=$(fin drush config-get "system.site" uuid --format=string)
    echo "SITE_UUID=${SITE_UUID}"  >> ../.docksal/docksal.env
else
    echo-green "SITE_UUID in docksal.env exists. Writing it to Database now."
    fin drupal config:override "system.site" uuid "${SITE_UUID}"
fi


echo -e "${green_bg} Step 6 ${NC}${yellow} Workaround for a Bug with Configs ...${NC}"
echo -e "${yellow} To make configs work existing shortcuts need to be removed and Module to be reinstalled. ${NC}"
cd ${DOCROOT_PATH}
fin drush ev '\Drupal::entityManager()->getStorage("shortcut_set")->load("default")->delete();'
fin rem shortcut


# Activate our fine config management module
echo -e "${green_bg} Step 7 ${NC}${green} Enable Config Split...${NC}"
cd ${DOCROOT_PATH}
fin drush en config_split -y
fin cr


echo -en "${green_bg} ... DONE! ${NC} "
echo -e "Open ${yellow}http://${VIRTUAL_HOST}${NC} in your browser to verify the setup."
echo -e "Or login directly here:"
fin dr uli


#-------------------------- END: Execution --------------------------------
```

## Output of fin sysinfo:

```
███  OS & BASICS
Linux Nyx-Manjaro 4.8.17-1-MANJARO #1 SMP PREEMPT Mon Jan 9 10:24:58 UTC 2017 x86_64 GNU/Linux
NAME="Manjaro Linux"
ID=manjaro
PRETTY_NAME="Manjaro Linux"
ANSI_COLOR="1;32"
HOME_URL="http://www.manjaro.org/"
SUPPORT_URL="http://www.manjaro.org/"
BUG_REPORT_URL="http://bugs.manjaro.org/"
fin version: 1.0.2
Mode:  Native / Docker for Mac/Windows
███  DOCKER
DOCKER_HOST:	

Docker: Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64
 Experimental: false
███  DOCKER COMPOSE
Docker Compose: docker-compose version 1.11.2, build dfed245b
docker-py version: 2.1.0
CPython version: 3.6.0
OpenSSL version: OpenSSL 1.0.2k  26 Jan 2017
███  DOCKER: IMAGES
REPOSITORY                                      TAG                 IMAGE ID            CREATED             SIZE
mysql                                           5.5                 94689ff4bcb5        5 weeks ago         256MB
schickling/swagger-ui                           latest              0ca58ca887d4        6 weeks ago         72.4MB
busybox                                         latest              00f017a8c2a6        7 weeks ago         1.11MB
phpmyadmin/phpmyadmin                           latest              3dee224cadd3        7 weeks ago         102MB
phpdockerio/nginx                               latest              59ef992ab263        2 months ago        5.33MB
docksal/dns                                     1.0                 20c7a535479d        3 months ago        61.6MB
docksal/ssh-agent                               1.0                 148220cc033d        3 months ago        12.2MB
docksal/web                                     1.0-apache2.2       a9c03f8a5f0f        3 months ago        188MB
docksal/db                                      1.0-mysql-5.5       9fb8c0aed5fc        3 months ago        255MB
docksal/cli                                     1.0-php7            5bc490ae4d93        3 months ago        747MB
docksal/vhost-proxy                             1.0                 9acc7ecca0ab        3 months ago        164MB
mailhog/mailhog                                 latest              b1ce818de155        3 months ago        18.9MB
docksal/solr                                    4.x-stable          d3e3de86ec7c        4 months ago        546MB
blinkreaction/drupal-solr                       3.x-stable          bde45d55a9cc        18 months ago       474MB
███  DOCKER: CONTAINERS
CONTAINER ID        IMAGE                     COMMAND                  CREATED             STATUS              PORTS                                      NAMES
a6f2e824656a        docksal/ssh-agent:1.0     "/run.sh ssh-agent"      2 months ago        Up 39 seconds                                                  docksal-ssh-agent
e3135ed0ca05        docksal/dns:1.0           "/opt/entrypoint.s..."   2 months ago        Up 38 seconds       192.168.64.100:53->53/udp                  docksal-dns
ac01ecf420a6        docksal/vhost-proxy:1.0   "/usr/local/bin/st..."   2 months ago        Up 38 seconds       0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   docksal-vhost-proxy
███  VIRTUALBOX
5.1.20r114628
███  NETWORK INTERFACES
Name:            vboxnet1
GUID:            786f6276-656e-4174-8000-0a0027000001
DHCP:            Disabled
IPAddress:       192.168.88.1
NetworkMask:     255.255.255.0
IPV6Address:     
IPV6NetworkMaskPrefixLength: 0
HardwareAddress: 
MediumType:      Ethernet
Status:          Down
VBoxNetworkName: HostInterfaceNetworking-vboxnet1

Name:            vboxnet0
GUID:            786f6276-656e-4074-8000-0a0027000000
DHCP:            Disabled
IPAddress:       192.168.44.1
NetworkMask:     255.255.255.0
IPV6Address:     
IPV6NetworkMaskPrefixLength: 0
HardwareAddress
MediumType:      Ethernet
Status:          Down
VBoxNetworkName: HostInterfaceNetworking-vboxnet0
```
## fin config output
```
---------------------
COMPOSE_PROJECT_NAME: MY_PROJECTNAME
COMPOSE_PROJECT_NAME_SAFE: MY_PROJECTNAME
COMPOSE_FILE:
/home/nm/.docksal/stacks/volumes-bind.yml
/home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME/.docksal/docksal.yml
ENV_FILE:
/home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME/.docksal/docksal.env

PROJECT_ROOT: /home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME
DOCROOT: web
VIRTUAL_HOST: MY_PROJECTNAME.docksal
IP: 192.168.64.100
MYSQL:

Docker Compose configuration
---------------------
networks: {}
services:
  cli:
    environment:
      PHP_IDE_CONFIG: serverName=MY_PROJECTNAME.docksal
      XDEBUG_CONFIG: idekey=PHPSTORM remote_host=192.168.10.1
      XDEBUG_ENABLED: '1'
    hostname: cli
    image: docksal/cli:1.0-php7
    volumes:
    - host_home:/.home:ro
    - docksal_ssh_agent:/.ssh-agent:ro
    - project_root:/var/www:rw
  clifast:
    hostname: cli
    image: docksal/cli:1.0-php7
    volumes:
    - host_home:/.home:ro
    - docksal_ssh_agent:/.ssh-agent:ro
    - project_root:/var/www:rw
  db:
    environment:
      MYSQL_DATABASE: default
      MYSQL_PASSWORD: user
      MYSQL_ROOT_PASSWORD: root
      MYSQL_USER: user
    hostname: db
    image: docksal/db:1.0-mysql-5.5
    ports:
    - '3306'
  pma:
    depends_on:
      db:
        condition: service_started
    hostname: pma
    image: phpmyadmin/phpmyadmin:latest
    labels:
      io.docksal.virtual-host: pma-MY_PROJECTNAME.docksal
  solr:
    hostname: solr
    image: docksal/solr:4.x-stable
    volumes:
    - /home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME/config/solr:/var/lib/solr/conf:rw
  web:
    depends_on:
      cli:
        condition: service_started
    environment:
      APACHE_DOCUMENTROOT: /var/www/web
      VIRTUAL_HOST: MY_PROJECTNAME.docksal
    hostname: web
    image: docksal/web:1.0-apache2.2
    labels:
      io.docksal.project-root: /home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME
      io.docksal.virtual-host: MY_PROJECTNAME.docksal
    volumes:
    - project_root:/var/www:ro
version: '2.1'
volumes:
  docksal_ssh_agent:
    external: true
  host_home:
    driver: local
    driver_opts:
      device: /home/nm
      o: bind
      type: none
  project_root:
    driver: local
    driver_opts:
      device: /home/nm/Code/MY_COMPANYNAME/MY_PROJECTNAME
      o: bind
      type: none

```

**Comment by lmakarov** [comment]
@drupallerina are you saying you are seeing this happening consistently now?

This happened to me once and I was (thankfully, I guess) not able to reproduce it since then.
At that time I thought it was some edge case / bug with removing docker volumes.

We do mount the host home directory as a volume in each project stack. In the config above it is this:

```
  host_home:
    driver: local
    driver_opts:
      device: /home/nm
      o: bind
      type: none
```

The volume is then mounted as READ-ONLY (changed after I had my own home folder wiped...) in cli:

```
    volumes:
    - host_home:/.home:ro
```

Even if it was mounted as read/write, docker should not be deleting the directory a bind mounted volume is pointing to.

BTW I see you are Docksal power-user based on the configuration and custom commands you have in place ;)

**Comment by achekulaev** [comment]
@drupallerina interesting setup you got!

1. I would recommend updating fin as you got 1.0.2 which is pretty old version

2. Could you specify when exactly you see files disappearing during your `fin init`? Maybe you can share output? As it's hard to tell when it's happening.

3. Can you share contents of `.docksal/commands/rem`?

**Comment by achekulaev** [comment]
Hi there. Any updates?

**Comment by drupallerina** [comment]
Sorry for letting you wait! It's going to take me a few days before I can test properly again, will post an update over the weekend. Thank you for your patience and for investigating along with me :)

**Comment by achekulaev** [comment]
Friendly reminder that we look forward your update in home to either locate a bug in our code or to locate a set of conditions that leads to the behavior.

**Comment by drupallerina** [comment]
Hello and thank you for your patience! 
@achekulaev 1. When I originally reported this issue I was running the dev branch. The output you see is from after restoring my system and downgrading to the master branch.
2. Files started disappearing right after stopping and removing the containers.
3. fin rem is just an alias for drush dre
@lmakarov thanks for the hint! It's probably related to a bug with removing docker volumes. I found two possible explanations for this: https://github.com/moby/moby/issues/32613 and https://github.com/moby/moby/pull/32327

Is there another reason we are mounting host/home in the container besides the forwarding of the ssh keys? https://github.com/moby/moby/issues/6396

**Comment by achekulaev** [comment]
> downgrading to the master branch

fin version in master branch is 1.6.0 not 1.0.2. But that version is I guess actually in line with the bug you have found in the docker queue. moby/moby#32613

If I'm reading it right, looks like in some cases volumes created with old versions of docker will get erased by 17.04 along with their data, which should not happen.

Looks like a workaround for anyone meeting this issue at this moment would be: `fin vm remove` and `fin vm start`. I guess re-creating volumes with newer version of docker should help.

**Comment by achekulaev** [comment]
If anyone will experience that please comment. Closing for now as most like it was a Docker bug that was addressed.

**Comment by Nebel54** [comment]
Liebe Kunden, Freunde und Spammer,
wir sind bis 11.6.2017 nicht im Büro und können Ihre Anfragen daher nicht sofort beantworten. Wir werden unsere E-Mails aber in unregelmäßigen Abständen abrufen und beantworten. 

Liebe Grüße
Oliver Köhler

**Comment by tjb328** [comment]
I experienced this yesterday on OSX 10.12.5. 
I was prompted:
``` 
UPDATE AVAILABLE 
fin [ 1.9.0 --> 1.10.0 ]
```

so I ran `fin update`. It took a really long time and I soon realized that my entire home folder was deleted. I did experience data loss one time before, but it only happened to my project folder.

Fortunately I had a time machine backup. I will try to reproduce this error when I have some time.

**Comment by lmakarov** [comment]
According to moby/moby#32613 the issue happens with volumes created on Docker Engine v1.12.6 and below. For Docksal that means anything before the [v1.1.0](https://github.com/docksal/docksal/releases/tag/v1.1.0) release (corresponding fin version 1.1.0), which went out on March 7, 2017.

You would need to have some project stack initialized before that date still laying around. Running a the `fin reset` command on that project will presumable result in the host home directory blow up.

Docker fixed the bug in 17.06.0, however the latest stable release of Docksal is using 17.04.0.
We are preparing a new release within a week, which requires Docker 17.06.0, where this bug was supposedly fixed.

**Comment by lmakarov** [comment]
Docksal [v1.4.0](https://github.com/docksal/docksal/releases/tag/v1.4.0) has been release yesterday and bundles Docker v17.06.0. This should now be resolved.
